### PR TITLE
feat: support updates for windows 14.0

### PR DESCRIPTION
**Pull request by mcdurdin**
This adds an endpoint for update checks for the bootstrapper, and later Keyman itself, for Keyman 14.0 and later versions. It has diverged enough from earlier API versions that a (partial) rewrite is acceptable.

Host references have been rewritten to use the standard `KeymanHosts` class. There may still be some outstanding references in legacy APIs.

I anticipate some more tidy-up on the updates API as the Keyman for Windows online update process matures.

## Pull Request Set

This is part of a set of five Pull Requests. The recommended order of review is:

* keymanapp/downloads.keyman.com#29 (basic infrastructure to support custom filenames for Keyman Windows Bootstrap)
* keymanapp/api.keyman.com#83 (this PR -- APIs for the Windows bootstrap)
* keymanapp/keyman.com#120 (universal link landing page + Windows)
* keymanapp/keyman#3304 (new Windows bootstrap)
* keymanapp/help.keyman.com#188 (documentation for Keyman Engine API changes)

### CODE REVIEW

File: script/windows/14.0/update/WindowsUpdateCheck.php [hunk 1]
```diff
@@ -1,15 +1,14 @@
 <?php
+  namespace Keyman\Site\com\keyman\api;
+
   require_once __DIR__ . '/../../../../tools/util.php';
 
-  class OnlineUpdate {
+  class WindowsUpdateCheck {
     const MSI_REGEX = '/^keyman(desktop)?\.msi$/';
     const BOOTSTRAP_REGEX = '/^setup\.exe$/';
     const BUNDLE_REGEX = '/^keyman(desktop)?-.+\.exe/';
```
Comment (darcywong00) [inline script/windows/14.0/update/WindowsUpdateCheck.php:9]: Will the name change from "Keyman Desktop" --> "Keyman for Windows" mean the installer names change too? 
e.g. regex `keyman(windows)`?

File: script/windows/14.0/update/WindowsUpdateCheck.php [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        if(preg_match($regex, $file)) {
+          $filedata->url = get_site_url_downloads() . "/windows/$tier/{$filedata->version}/{$file}";
+          return $filedata;
+        }
+      }
+
+      return FALSE;
+    }
+
+    private function IsSameMajorVersion($v1, $v2) {
```
Comment (darcywong00) [inline script/windows/14.0/update/WindowsUpdateCheck.php:71]: This looks unused and maybe can be removed

File: script/windows/14.0/update/WindowsUpdateCheck.php [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      $KeyboardDownload->url = $this->BuildKeyboardDownloadPath($KeyboardDownload->id, $KeyboardDownload->version);
+      if($KeyboardDownload->url === FALSE) {
+        // Unable to build a url for the keyboard, would only happen if downloads.keyman.com was out of sync with
+        // api.keyman.com
+        return FALSE;
+      }
+      return $KeyboardDownload;
+    }
+
+    private function BuildKeyboardDownloadPath($id, $version) {
```
Comment (darcywong00) [inline script/windows/14.0/update/WindowsUpdateCheck.php:150]: Is `$version` needed?

File: script/windows/14.0/update/WindowsUpdateCheck.php (hunk 1 above)
Comment (mcdurdin) [reply script/windows/14.0/update/WindowsUpdateCheck.php:13]: I am proposing that the new filename for a bundled install be `keyman-a.b.c.exe`.

File: script/windows/14.0/update/WindowsUpdateCheck.php (hunk 2 above)
Comment (mcdurdin) [reply script/windows/14.0/update/WindowsUpdateCheck.php:71]: Noted for cleanup in follow-up PR.

File: script/windows/14.0/update/WindowsUpdateCheck.php (hunk 3 above)
Comment (mcdurdin) [reply script/windows/14.0/update/WindowsUpdateCheck.php:151]: Noted for cleanup in follow-up PR.

File: script/windows/14.0/update/WindowsUpdateCheck.php [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        $keyboard = $this->BuildKeyboardResponse($id, $version, $appVersion);
+        if($keyboard !== FALSE) {
+          $keyboards[$id] = $keyboard;
+        }
+      }
+      return $keyboards;
+    }
+
+    private function BuildKeyboardResponse($id, $version, $appVersion) {
+      // TODO: this should use the class instead of file_get_contents
```
Comment (jahorton) [inline script/windows/14.0/update/WindowsUpdateCheck.php:94]: I assume the `// TODO:` is for later work?

File: tests/WindowsUpdateCheckTest.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+final class WindowsUpdateCheckTest extends ApiTestCase
+{
+  private const SchemaFilename = "/windows-update/14.0/windows-update.json";
+
+  public function testSimpleResultValidatesAgainstSchema(): void
+  {
+    $schema = TestUtils::LoadJSONSchema(self::SchemaFilename);
+
+    $u = new \Keyman\Site\com\keyman\api\WindowsUpdateCheck();
+    $json = $u->execute('alpha', '14.0.100', ['khmer_angkor'=>'1.0.2', 'foo'=>'0.1']);
```
Comment (jahorton) [inline tests/WindowsUpdateCheckTest.php:21]: From this, I take it that this parallels an in-app "check for updates" menu item?  Said menu item then checks to update both Keyman for Windows and any resources that are installed?

File: script/windows/14.0/update/WindowsUpdateCheck.php (hunk 4 above)
Comment (mcdurdin) [reply script/windows/14.0/update/WindowsUpdateCheck.php:94]: Yes. It's not urgent, but an unnecessary http call to the same host seems wasteful.

File: tests/WindowsUpdateCheckTest.php (hunk 5 above)
Comment (mcdurdin) [reply tests/WindowsUpdateCheckTest.php:21]: At present, this API endpoint is used by the setup bootstrap program, which looks for latest Keyman Desktop as well as keyboards specified by parameter. While the Keyman Desktop online update check will move to use this as well, I have not yet checked the code around that.

**Pull request merged**